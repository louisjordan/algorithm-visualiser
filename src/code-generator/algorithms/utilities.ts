import * as fs from 'fs';
import * as path from 'path';

const algorithmsRoot = path.join(__dirname, '../../algorithms/');

export function algorithmPaths() {
    return fs
        .readdirSync(algorithmsRoot)
        .map((algorithm: string) => ({
            algorithm,
            algorithmPath: path.join(algorithmsRoot, algorithm),
        }))
        .filter((path) => fs.statSync(path.algorithmPath).isDirectory());
}
