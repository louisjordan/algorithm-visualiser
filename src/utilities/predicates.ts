// is object predicate
export function isObject(v: unknown): v is object {
    return typeof v === 'object' && v !== null;
}

export function isArray(v: unknown): v is Array<unknown> {
    return Array.isArray(v);
}

export function isFunction(v: unknown): v is Function {
    return typeof v === 'function';
}
