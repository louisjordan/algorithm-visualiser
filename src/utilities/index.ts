import { isObject } from './predicates';

// re-export predicates
export * from './predicates';

// recursively freeze an object
export function deepFreeze<T extends object>(obj: T): T {
    return Object.freeze(
        Object.keys(obj).reduce((obj, k) => {
            const key = k as keyof T;
            const value = obj[key];
            if (isObject(value)) {
                obj[key] = deepFreeze(value);
            } else {
                obj[key] = value;
            }
            return obj;
        }, obj)
    );
}

export function uuid(length: number = 8): string {
    if (length < 1) {
        throw new Error('Slug must be at least one character');
    } else if (length === Infinity) {
        throw new Error("We haven't got all day");
    }

    let slug = '';

    while (slug.length < length) {
        slug += Math.random()
            .toString(36)
            .slice(2, -1);
    }

    return slug.substr(0, length);
}
