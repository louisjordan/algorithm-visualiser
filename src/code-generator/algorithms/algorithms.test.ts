import * as path from 'path';
import { exec } from 'child_process';
import { algorithmPaths } from './utilities';

describe('Algorithm code generator', () => {
    describe('matches snapshots', () => {
        beforeAll(() => {
            exec('npm run build-sources');
        });

        algorithmPaths().forEach(({ algorithm, algorithmPath }) => {
            it(`should match ${algorithm} snapshot`, () => {
                expect(
                    JSON.stringify(
                        require(path.join(algorithmPath, 'source.json'))
                    )
                ).toMatchSnapshot();
            });
        });
    });
});
