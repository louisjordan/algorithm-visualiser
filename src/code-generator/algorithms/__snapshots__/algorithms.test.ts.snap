// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Algorithm code generator matches snapshots should match graph-depth-first-search snapshot 1`] = `"{\\"functions\\":[\\"function graphDepthFirstSearchRecursive(\\\\n    graph: DFSGraph,\\\\n    start: DFSGraphNode\\\\n): DFSGraph {\\\\n    start.value.visited = true;\\\\n\\\\n    for (let node of start.neighbours()) {\\\\n        if (!node.value.visited) {\\\\n            graphDepthFirstSearchRecursive(graph, node);\\\\n        }\\\\n    }\\\\n\\\\n    return graph;\\\\n}\\"]}"`;

exports[`Algorithm code generator matches snapshots should match insertion-sort snapshot 1`] = `"{\\"functions\\":[\\"function insertionSort(list: number[]) {\\\\n    let position = null;\\\\n    let pointer = null;\\\\n\\\\n    position = 1;\\\\n\\\\n    while (position < list.length) {\\\\n        let x = list[position];\\\\n\\\\n        pointer = position - 1;\\\\n\\\\n        while (pointer >= 0 && list[pointer] > x) {\\\\n            list[pointer + 1] = list[pointer];\\\\n            pointer--;\\\\n        }\\\\n\\\\n        list[pointer + 1] = x;\\\\n\\\\n        position++;\\\\n    }\\\\n\\\\n    return list;\\\\n}\\"]}"`;

exports[`Algorithm code generator matches snapshots should match selection-sort snapshot 1`] = `"{\\"functions\\":[\\"function selectionSort(list: number[]) {\\\\n    for (let position = 0; position < list.length - 1; position++) {\\\\n        let smallest = position;\\\\n\\\\n        for (let pointer = position + 1; pointer < list.length; pointer++) {\\\\n            if (list[smallest] > list[pointer]) {\\\\n                smallest = pointer;\\\\n            }\\\\n        }\\\\n\\\\n        let x = list[position];\\\\n        list[position] = list[smallest];\\\\n        list[smallest] = x;\\\\n    }\\\\n\\\\n    return list;\\\\n}\\"]}"`;
